### DISCUSSION TITLE
Feature proposal: adding Edits to Kitsu

### DISCUSSION THREAD

**Discussion by ex5** [cgwire/kitsu#801] [Show and tell]
It would be very useful to have Edits in Kitsu to be able to cooperate on the final result of a production, leveraging all the powerful functionality Kitsu already provides, such as revisions, comments, statuses, video previews and so on.

For this purpose, I'd like to propose adding an Edit Entity, Edit task type and the corresponding menu section and pages.
Here, Edit task type is enabled in an existing production, which makes Edits menu section appear:

https://user-images.githubusercontent.com/176934/150321758-20d6ca87-578a-4090-81ea-814c29bf62c6.mp4

In terms of data structure and presentation in the first iteration of the implementation, Edits are very similar to Assets.
They can be created in a similar manner, and previews can be added to an Edit task to create revisions/versions of a single Edit:

https://user-images.githubusercontent.com/176934/150321649-765b13a3-ecb7-48a3-ab96-639a58b6771c.mp4

https://user-images.githubusercontent.com/176934/150321543-b99bfe37-99b2-4e76-a772-239b60f6fe09.mp4

However, in the second iteration of the front-end implementation, Edit task page should become more  accessible, e.g. clicking on the Edit would display a page similar to a TaskPage, but more focused on video playback and versions.
Below is a mockup of how the Edit page could look like:

![EditPageMockup](https://user-images.githubusercontent.com/176934/150325228-4d88eca7-961e-43be-9b61-499a867b2b5f.png)

<hr>

For the first iteration of the implementation (everything shown in the 3 videos) I already have patches mostly ready for review.

Before submitting the PRs however (and before diving into the implementation of the Edit page), I'd like to try and hash out any potential issues with Edits here.

If what it shown in the videos raises no objections, I'd like to move forward with PRs for Edit entity/task type and the corresponding front-end changes.

**Comment by frankrousseau** [comment]
Thank you a lot @ex5 for this proposition! I will discuss this with @gwendupre. On my side, from what I see it looks good to me. I'm just wondering about three things:

* Does edit make sense only for the short productions?
* How does it differ from a playlist build?
* The layout for the task page may work for other tasks too. Should we apply it to all tasks?

**Reply by ex5** [reply]
> Does edit make sense only for the short productions?

The edits make sense for all productions, I think.
In case of series/show edits are linked to episodes, similarly to assets.

![2022-01-25-101647_1026x413_scrot](https://user-images.githubusercontent.com/176934/150948059-80fdf113-59c8-42fe-9ebd-6556a01b0635.png)


> How does it differ from a playlist build?

Edits differ from playlists conceptually: they are an externally created final(-ish) results of the whole production, and, as such, they can differ from any playlist in a variety of ways, e.g. sound effects, music, temporary or otherwise, cuts and so on.
In addition to the "Edit" task, edits might also require a unique set of additional tasks, such as sound or DCP.

> The layout for the task page may work for other tasks too. Should we apply it to all tasks?

Sounds good to me :+1:

**Comment by flablog** [comment]
Hi there !

This looks interesting, but I was wondering what you consider an Edit ? What is the "level" here, the whole movie/episode for example ? or a sequence within a larger edit ?
Then you use the department revision system to upload upgraded versions of that edit ?
Or you create one entity for each of your edits, edit_v01, etc ?

And last question, could this be a good place to save some important metadata such as the EDL, or opentimeline.io information, with upgraded and easy to spot changes in the edit ?
Like from this version to this version shot 3 as been cut 10 frames, etc.
Which is always a nightmare to handle well ?

**Reply by ex5** [reply]
> This looks interesting, but I was wondering what you consider an Edit ? What is the "level" here, the whole movie/episode for example ? or a sequence within a larger edit ?

In this context, the "final result" of a production is considered an edit, so, yes, the whole movie or a whole episode.

> Then you use the department revision system to upload upgraded versions of that edit ?
> Or you create one entity for each of your edits, edit_v01, etc ?

The former: it would make sense to create a single Edit entity and then use existing revision system (on the third video it should be visible that there are 2 revisions of the edit, multiple revisions are also shown at the mockup image of the Edit page).

Multiple Edit entities make sense for series/show --- one per each episode, or in case when one edit is conceptually different from another (like editor/director/final cuts being revisioned separately maybe?), but I'd guess this should be up to a particular studio's workflow.

> And last question, could this be a good place to save some important metadata such as the EDL, or opentimeline.io information, with upgraded and easy to spot changes in the edit ?
> Like from this version to this version shot 3 as been cut 10 frames, etc.
> Which is always a nightmare to handle well ?

I'm not sure this would be a good place to store EDL metadata in a usable way, unfortunately, but storing it as an attachment is possible and adding some of the metadata such as start/end frame to the entity is also possible without any additional changes to Kitsu:
 
![2022-01-25-104551_3440x1440_scrot_](https://user-images.githubusercontent.com/176934/150952665-ec441581-46c8-4f44-8e24-44e95b7eee25.png)

**Comment by Ramel** [comment]
Is there a relation between Edit and Playlist? Is the starting point of an Edit is a validated/closed Plalylist build?

**Reply by ex5** [reply]
It can be, however I'd rather not make this assumption part of the implementation in any way, because whether an Edit starts as a playlist or not is up to a particular studio's workflow.

**Comment by ddesmond** [comment]
Suggestion: should be compatible to https://github.com/PixarAnimationStudios/OpenTimelineIO at least so we would be able to transfer back and forth various formats and apps, We for one would love to have this managed from Nuke studio or davinci resolve.

**Reply by fsiddi** [reply]
The idea for this feature is to simply visualize (and review) the output of the edit process, not the data itself. Pushing an edit export should be possible from any software which integrates with the Zou API. If you are interested in editorial data visualization you could check [Watchtower](https://studio.blender.org/blog/introducing-watchtower/), which can connect with Kitsu.

**Comment by frankrousseau** [comment]
This feature has been implemented by @ex5 and added to the main branch.
